extern templates

etienneschmitt opened the pull request.

etienneschmitt committed:
first draft of extern templates

Signed-off-by: Etienne Schmitt <etienne.schmitt@inria.fr>

etienneschmitt committed:
removed Eigen in core

Signed-off-by: Étienne Schmitt <etienne.schmitt@inria.fr>

etienneschmitt committed:
fixed extern templates

Signed-off-by: Étienne Schmitt <etienne.schmitt@inria.fr>

etienneschmitt committed:
Merge branch 'develop-schmitte' into extern_templates

# Conflicts:
#	cgogn/core/CMakeLists.txt
#	cgogn/core/map/cmap1.h
#	cgogn/core/map/cmap2.h
#	cgogn/core/map/cmap2_builder.h
#	cgogn/utils/name_types.h

Signed-off-by: Etienne Schmitt <etienne.schmitt@inria.fr>

etienneschmitt committed:
added some extern templates

Signed-off-by: Etienne Schmitt <etienne.schmitt@inria.fr>

etienneschmitt committed:
Merge branch 'develop-schmitte' into extern_templates

etienneschmitt committed:
more uses of extern template.

Signed-off-by: Étienne Schmitt <etienne.schmitt@inria.fr>

pierrekraemer commented:
Au passage, il faudrait voir du côté de LTO https://gcc.gnu.org/wiki/LinkTimeOptimization qui permet notamment d'inliner des fonctions lors du link.

etienneschmitt committed:
fixed compilation with MSVC

Signed-off-by: Etienne Schmitt <etienne.schmitt@inria.fr>

pierrekraemer commented:
Dans CMake, il y a apparemment une option nommée INTERPROCEDURAL_OPTIMIZATION qui active LTO sur les compilateurs où c'est disponible.
A voir si ça fonctionne correctement parce que d'après ce que j'ai compris, LTO permet d'optimiser le code généré au moment de linker plusieurs .o dans un binaire. Si ce binaire est un exécutable, c'est terminé. Si c'est une lib statique des optimisations sont appliquées à la génération de la lib, mais pour qu'elles puissent à nouveau avoir lieu au moment où on linke un exécutable utilisant cette lib, il faut que la lib ait été générée avec gcc-ar et le plugin LTO.
Exemple d'informations : http://stackoverflow.com/questions/25878407/how-can-i-use-lto-with-static-libraries

etienneschmitt commented:
Bon alors la documentation sur INTERPROCEDURAL_OPTIMIZATION est plus que laconique. Aucune idée de ce que ça fait mais avec un make VERBOSE=1 je ne vois aucun flag lto ni de gcc-ar...

etienneschmitt commented:
Je peux essayer de faire ça à l'ancienne. Mais je pense que l'utilisation du LTO doit rester optionnelle : il semble subsister quelques bugs...

pierrekraemer commented:
Ok. Mais du coup il faudrait quand même voir si faire des extern templates n'a pas un effet néfaste sur les performances en empêchant le compilateur d'inliner des méthodes des classes templates (chose a priori rendue à nouveau possible par LTO qui inline au moment du link).

etienneschmitt committed:
fixed compilation with clang.

Signed-off-by: Étienne Schmitt <etienne.schmitt@inria.fr>

etienneschmitt commented:
Ouais on comparera les perfs. En cas de besoin l'option peut se désactiver rapidement ! 
J'ai pas mal testé le LTO et rien que faire un truc fonctionnel sous clang n'est pas si évident.  Est-ce que c'est nécessaire pour que la PR soit acceptée ?

etienneschmitt committed:
fixed compilation with gcc ;)

Signed-off-by: Étienne Schmitt <etienne.schmitt@inria.fr>

pierrekraemer commented:
Ben non, si ça se désactive facilement c'est bon.
Est-ce que en l'état c'est activé par défaut ou non ?
Si ça coûte cher en performance, on le laissera désactivé par défaut. Mais on pourra le laisser quand même en stand-by pour le moment où on arrivera à faire fonctionner LTO.

etienneschmitt commented:
yep c'est désactivé par défaut. Et niveau perf j'ai pas vu de diff en mesurant les résultats du mini-bench de test_import ( que les caches soient activés ou non).

pierrekraemer commented:
Pourquoi c'est tout commenté dans CMap3 ?

etienneschmitt commented:
Parce que ça compile pas. :-1:

etienneschmitt commented:
tellement d'erreurs que mon qt creator plante ;-)

etienneschmitt committed:
fixed compilation of cmap3.

Note : each time a {dart,cell}Marker is constructed in a map, a good practice would be to give *this->to_concrete() instead of this as argument.

Signed-off-by: Étienne Schmitt <etienne.schmitt@inria.fr>

etienneschmitt commented:
haha j'ai même de nouveau réussi à faire planter gcc !!!
Mais le problème est maintenant résolu. Il faut bien faire attention lors de la création d'un marker, il ne faut pas lui donner "this" lors de la construction mais plutôt le cast vers la carte concrete.

david-cazier commented:
A la fin du fichier cmap3.h, est-ce qu'il ne faudrait pas plutôt inclure le cmap3.hpp au lieu de dupliquer les lignes de codes du type extern template ... ;
Il me semble qu'il y a un risque de désynchronisation des 2 fichiers.

david-cazier commented:
en ajoutant externe via un #define

etienneschmitt commented:
quoi ? un .hpp ? ....

etienneschmitt commented:
dans un hpp ça ne rime à rien puisqu'on veut explicitement instancier une classe. Il faut que ce soit dans un cpp. Et il faut que les extern apparaissent dans les headers...

pierrekraemer commented:
Je ne comprends pas pourquoi il faut faire this->to_concrete() quand on construit un Dart/CellMarker dans une méthode de CMapX. Est-ce que this n'est pas déjà du bon type dans ce contexte ?

etienneschmitt commented:
dans Map3::foreach_xxx quand on appelle Inherit::foreach_yyyy (inherit = map2), si dans cette méthode on construit un marker, "this" sera de type map2\* tandis que le DartMarker sera de type cgogn::DartMarker< Map3 > (car ConcreteMap = Map3) ---> erreur de compil.

Edit : < Map3 > n'apparaissait pas

etienneschmitt commented:
Même si this réfère à un objet de type map3 dans le contexte c'est un pointeur vers une map2. On est obligé de le caster.

david-cazier commented:
faute de frappe .cpp pas .hpp, oui

en pratique a la fin du c3map.h on aurait include "c3map.cpp" avec #define
EXTERN qui va bien
ca évite d'écrire 2 fois les "extern tempate class ..." (une fois dans le
.h et une fois dans le .cpp)

_--_
_David Cazier_
Professeur à l'université de Strasbourg - Laboratoire ICube
Equipe projet INRIA MIMESIS

2016-01-14 13:44 GMT+01:00 Etienne Schmitt notifications@github.com:

> quoi ? un .hpp ? ....
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/cgogn/CGoGN_2/pull/66#issuecomment-171635034.

etienneschmitt commented:
Inclure des cpp dans les .h ?

etienneschmitt commented:
Dans le .cpp il n'y a pas d'extern.

david-cazier commented:
d'ou le #define EXTERN vide dans le cpp et définie à extern dans le .h

inclusion cpp dans .h => pourquoi pas

sinon il faut un 3ème fichier inclus dans les .h et .cpp (comme tu veux)

etienneschmitt commented:
Je pense que ma solution est mieux, pas lourde. Ce sont des lignes de code qui ne sont pas destinées à changer régulièrement.
Je vote pour accepter ma PR en l'état.

david-cazier commented:
:scream: pas besoin d'être lourd pour faire des erreurs, il suffit d'ajouter une ligne dans 6 mois et d'oublier de l'ajouter dans l'autre fichier. Mais on en a assez discuté. Fais comme tu veux.

etienneschmitt commented:
En fait si jamais ton cas se produit,  il y aura une erreur au link...

pierrekraemer commented:
Concernant les extern templates, je suis assez d'accord avec Etienne. Quand on écrit une définition d'une variable externe quelque part, on écrit la définition effective dans la foulée (sinon c'est qu'on ne sait pas vraiment ce qu'on fait..).
En revanche, faire des include de .cpp avec des define me semble très casse gueule pour le jour où il y aura autre chose que ça dans le .cpp.

etienneschmitt commented:
(sous windows en tout cas, l'erreur au link).
Pierre et concernant le to_concrete() tu n'es tjours pas convaincu ? j'ai pourtant passé du temps à lire les msgs d'erreur de clang et gcc ;)

pierrekraemer commented:
Sinon pour les Dart/CellMarker et les histoires de to_concrete.
Ca me semble lourd de devoir les instancier en castant dans le type concret.
Actuellement les types Dart/CellMarker sont définis dans chaque classe CMapX par référence à la définition de la classe mère, jusqu'à MapBase qui les définit avec le type concret en paramètre.
On pourrait tout à fait définir ces types dans chaque CMapX avec le type Self de la CMapX concernée. De toute manière, on ne s'attend pas, dans une CMap2, qu'un DartMarker soit capable de marquer une orbite qui implique un PHI3. Donc le foreach_dart_of_orbit appelé peut bien être celui de CMap2, même dans le cas où le type concret est une CMap3.
Je ne sais pas si je suis clair..
De toute manière je projète de refaire une passe sur les méthode foreach_dart_of_xxx des classes CMapX pour les rendre plus lisible en lien avec le nom des orbites (PHI1, PHI2, PHI1_PHI2, etc..).

etienneschmitt commented:
Avec Self ça donnerait exactement le même bug !

pierrekraemer commented:
Bon on va accepter en l'état et je verrai par la suite.

pierrekraemer commented:
Mais sinon, comme à mon avis, cette PR et #67 ne sont pas compatibles, on le fait dans quel ordre ?

etienneschmitt commented:
On va prendre celle-ci avant, étant donné que Lionel est en train de terminer la sienne. Je serai avec lui pour faire le merge avant que la sienne ne soit acceptée ;)

pierrekraemer commented:
Ok ça roule :-)
